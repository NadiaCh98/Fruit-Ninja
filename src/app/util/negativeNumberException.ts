export class NegativeNumberException extends Error {
  constructor() {
    super('Value can not be negative');
  }
}
