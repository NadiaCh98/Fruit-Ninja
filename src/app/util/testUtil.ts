import { FruitSequence } from './../game/models/fruitData';
import { FruitsGenerator } from '../game/services/fruitsGenerator';

export const getFruitsSeriesByGeneratorAndIteraction = (
  generator: FruitsGenerator,
  iteraction: number
): FruitSequence => {
  return generator.generateFruitsSequence(iteraction);
};
