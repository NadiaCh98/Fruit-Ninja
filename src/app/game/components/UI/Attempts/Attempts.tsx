import React from 'react';
import { times } from 'lodash';
import { RIGHT_POSITION } from '../../../../common/constant';
import { GameControls } from '../../GameControls/GameControls';
import { ActiveAttempt, CanceledAttempt } from '../Icons/Icons';
import { NegativeNumberException } from '../../../../util/negativeNumberException';

interface AttemptsProps {
  readonly totalAttempts: number;
  readonly remainingAttempts: number;
}

export const Attempts: React.FC<AttemptsProps> = React.memo(
  ({ totalAttempts, remainingAttempts }) => {
    if (remainingAttempts > totalAttempts) {
      throw new Error('Total attemps can not be less than remaining');
    }
    if (remainingAttempts < 0 || totalAttempts < 0) {
      throw new NegativeNumberException();
    }
    const attempts = [
      ...times(totalAttempts - remainingAttempts).map((_, i) => (
        <CanceledAttempt key={i} />
      )),
      ...times(remainingAttempts).map((_, i) => (
        <ActiveAttempt key={totalAttempts + i} />
      )),
    ];
    return <GameControls style={RIGHT_POSITION}>{attempts}</GameControls>;
  }
);

Attempts.displayName = 'Attemps';
