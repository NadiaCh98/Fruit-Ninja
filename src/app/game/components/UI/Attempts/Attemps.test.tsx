import React from 'react';
import { render } from '@testing-library/react';
import { Attempts } from './Attempts';
import { NegativeNumberException } from '../../../../util/negativeNumberException';

describe('Attempts', () => {
  const attempts = 3;
  it('has a correct template #1', () => {
    const { container } = render(
      <Attempts totalAttempts={attempts} remainingAttempts={0} />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('has a correct template #2', () => {
    const { container } = render(
      <Attempts totalAttempts={attempts} remainingAttempts={1} />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('has a correct template #3', () => {
    const { container } = render(
      <Attempts totalAttempts={attempts} remainingAttempts={2} />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('has a correct template #3', () => {
    const { container } = render(
      <Attempts totalAttempts={attempts} remainingAttempts={3} />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('asserts that should throw error if total attemps < remaining attemps', () => {
    expect(() =>
      render(<Attempts totalAttempts={attempts} remainingAttempts={5} />)
    ).toThrowError('Total attemps can not be less than remaining');
  });

  it('asserts that should throw error if remaining attemps < 0', () => {
    expect(() =>
      render(<Attempts totalAttempts={attempts} remainingAttempts={-1} />)
    ).toThrowError(NegativeNumberException);
  });

  it('asserts that should throw error if total attemps < 0', () => {
    expect(() =>
      render(<Attempts totalAttempts={-3} remainingAttempts={-7} />)
    ).toThrowError(NegativeNumberException);
  });
});
