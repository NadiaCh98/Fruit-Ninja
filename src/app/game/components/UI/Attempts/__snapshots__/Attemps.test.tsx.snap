// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Attempts has a correct template #1 1`] = `
<div
  class="sc-bdnxRM lhbDim"
  style="top: 0px; right: 0px;"
>
  <svg>
    canceledAttempt.svg
  </svg>
  <svg>
    canceledAttempt.svg
  </svg>
  <svg>
    canceledAttempt.svg
  </svg>
</div>
`;

exports[`Attempts has a correct template #2 1`] = `
<div
  class="sc-bdnxRM lhbDim"
  style="top: 0px; right: 0px;"
>
  <svg>
    canceledAttempt.svg
  </svg>
  <svg>
    canceledAttempt.svg
  </svg>
  <svg>
    activeAttempt.svg
  </svg>
</div>
`;

exports[`Attempts has a correct template #3 1`] = `
<div
  class="sc-bdnxRM lhbDim"
  style="top: 0px; right: 0px;"
>
  <svg>
    canceledAttempt.svg
  </svg>
  <svg>
    activeAttempt.svg
  </svg>
  <svg>
    activeAttempt.svg
  </svg>
</div>
`;

exports[`Attempts has a correct template #3 2`] = `
<div
  class="sc-bdnxRM lhbDim"
  style="top: 0px; right: 0px;"
>
  <svg>
    activeAttempt.svg
  </svg>
  <svg>
    activeAttempt.svg
  </svg>
  <svg>
    activeAttempt.svg
  </svg>
</div>
`;
